# Invetigating Adaptive Epistasis using *Dmel* RILs

This is a collaborative project involving genotyping and phenotyping of two sets of *Drosophila melanogaster* RILs. One set consists of a France *vs* Zambia (France RILs) cross and the other an Ethiopia *vs* Zambia (Ethiopia RILs) cross. For each set, two phenotypes have been measured. For the France RILs, the phenotypes were ethanol tolerance and cold tolerance. For the Ethiopia RILs, the phenotypes were wing length and pigmentation. The pigmentation phenotype is further subdivided into three distinct measures: pigmentation score in gray scale for the background of the mesopleura (a thorax trait), and two abdominal traits, a pigmentation score the 4th abdominal segment (A4 Background), and the proportion of the 4th abdominal background covered by a black stripe.

The genomes for each RIL and the parental strains were obtained with Next Gen sequencing using Illumina platform.

## DNA Alignment
*Scripts for this section are located in the _Alignment_ directory.*

This section contains the steps taken to analyse the genomic data, starting with the fastq files output from the Illumina sequecing.

We used the UW-Madison Center for High Throughput Computing (CHTC) to perform the DNA alignment of our samples. Using CHTC allowed us to greatly parallelize the alignment step. It required that our initial fastq files be split in shorter files and each of these shorter fastq files were submitted as independent jobs to be aligned - and then latter merged back together. The *fastq_processing1.sh* handles the splitting step for each sample and generates a file listing all the shorter fastq blocks which will be used for the parallel submission. After running this script locally, in the same folder with the fastq files to be split, the block files and the newly generated *Seq_list_single.txt* file need to be uploaded to CHTC.

On CHTC, after the files are placed in their appropriate directories, we submitted the jobs with the *align_pt1.sub* script. This script uses as executable file the *align_pt1.sh* script. The *align_pt1.sh* script will perform the DNA alignment using bwa mem. In the end, we have bam files each fastq block. The block files need to be download to continue the pipeline locally.

The complete pipeline used in each parallel job also included steps using *samtools* to process the bam files. It uses *collate* to organize the reads, *fixmate* to correct any flaws in read-pairing, *sort* to sort the reads and finally *markdup* with *-r* flag to identify and remove duplicate reads. The next step uses a simple script, *block2fasta.sh* to merge the blocks and remove duplicates. To remove duplicates, it uses *samtools* to process the merged bam file. The merged bam file is processed using *collate* to organize the reads, *fixmate* to correct any flaws in read-pairing, *sort* to sort the reads and finally *markdup* with *-r* flag to identify and remove duplicate reads. if any was still left after the merging. Then, it uses *Picard* and *GATK* (with specific versions and requirements, see script) to re-align the bam file around indels and call indels and SNPs. It uses *bcftools* and custom scripts (*indel_vcf.py*, *split_fasta.py*, and *iupac_fasta.py*) to generate fasta files with ambiguity code for each *Dmel* chromosome arm, it will filter out (replace with N) sites with low quality (QUAL<32), missing data, indels and 3 bp surrounding indels. 

The *block2fasta.sh* script will create a directory to store all the intermediary files, in case you need to check them or easily delete all of them, and will also generate a log file to ease potential need to troubleshoot.

The final bam file can be used on the RILs downstream ancestry analysis. The final fasta (actually unecessary for the rils, could be deleted) can be used for downstream parental genome analysis. The sites.vcf file is being kept because it had been needed in previous pipelines from the Lab. I will reconsider if it should be moved to the intermediary files directory.

## Ancestry calls
*Scripts for this section can be found in the _Ancestry_ directory.*

Panel files were obtained with the script _panel_pipeline.sh_. This script uses the perl script _find\_snps.pl_ to generate one panel file per chromosome arm for each RIL data set. Order of the genomes matter and herein EF and FR were used first and will have genotype 0. The panel files are stored in their RIL's respective folders along the bam files for all the RILs (already without duplicates and realigned around indels). Then, we call the script _ancestry.sh_ to use the panels and bam files to call AncestryHMM and obtain posterior ancestry probabilities for each RIL. The _ancestry.sh_ script loops through each RIL and each chromosome arm to obtain ril-arm-specific mpileup files (obtained with _samtools mpileup_) and the ril-arm-specific panel files (obtained with the _populate\_snp\_matrix.pl_ script). To adjust the distance between the SNPs accordingly to the RIL experimental design we used _recomb.csv_ files generated for each RIL set and chromosome arm based on how many generations of recombination they had and on the recombination rates for _Dmel_ from Comeron et al. 2012 (FR RILs are F12 and EF RILs are F13. Therefore, recombination rates were multiplied by 11 and 12, respectively). The distance adjustment is made with the _rec\_map.py_ script. With the panel files per RIL per arm we call _ancestry\_hmm_ to obtain the posterior probabilities.

Following this, we used the script _RIL\_genotypes\_windows.pl_, placed in the _posteriors_ folder, to call ancestries for windows and not per SNP. The files delimiting the window sizes need to be located in the _posterior_ folders as well. Here, we used _windows\_ZI1k\_Chr*.txt_, which has windows defined to contain 1k SNPs in the ZI population. We have to manually update the genotypes perl script to ensure the script has the correct RILs at the _"@ my RILS"_ variable, we also need to manually input the output file name.

We obtained the ancestry calls in the files: _*\_RIL\_genotypes\_winZI1k.txt_.

## G x P input files
*Scripts for this section can be found in the _GenoPheno_ directory.*

We combined the ancestry genotypes per window obtained above with phenotypic data to generate files that will be analyzed in our mapping and epistasis analyses. Currently, our focus is on using the r/qtl R package (source). We used a set of files and scripts to generate the input files, as described below.

We used _input\_rqtl\_nopheno.py_ to transform the data into the input format for rqtl, then used _input\_rqtl\_addpheno.py_ to add the desired phenotype. We used the script _missing\_genotypes.py_ to attempt to imputate some of the missing windows (if they were surrounded by the same genotype and there were no more than 10 contiguous missing windows). We used the script _window\_merge\_genotype.py_ to merge neighboring windows if the genotype for these windows were the same for all the RILs. Then, we used the script _comeron\_rqtl.py_ to add recombination data to the file, basically adding the cM position for the midpoint of each window. Lastly, we used the script __ to split the chromosomes 2 and 3 in three parts with equal number of windows (instead of the default 2L and 2R, 3L and 3R). The split is done to reduce the number of markers per chromosome label, otherwise r/QTL would result in an error trying to infer missing data - which is done considering the information from all the markers in a chromosome. To ensure that we still include nearby windows in case there is missing data near the boundaries in which the chromosomes were split we added 50 windows to the extremities, which overlap with windows from other segments. However, the added windows are labeled as additions (_\_ext_) and results from these windows are ignored. We also added to this github directory an example bash script with a pipeline from the ancestry call output from the previous step all the way to the input file for r/qtl.

## R/qtl
*Scripts for this section can be found in the _rqtl_ directory.*
We used _window\_merge\_genotype.py_ to reduce the input file combining windows there were next to each other and had the same genotype for ALL the lines used, since the tests would be identical.
